libsel4vmmplatsupport/src/drivers/virtio_block_emul.c

```diff
@@ -0,0 +1,216 @@
+            if (i == 1) {
+                buf_len = desc.len;
+            }
+            i++;
+            len += this_len;
+            desc_idx = desc.next;
+        } while (desc.flags & VRING_DESC_F_NEXT);
+        /* ship it */
+        emul_tx_cookie_t *cookie = calloc(1, sizeof(*cookie));
+        assert(cookie);
+        cookie->desc_head = desc_head;
+        cookie->vaddr = vaddr;
+
+        /* Currently we can only handle buffers of a certain size or less.
+         *  We could fix this, but not sure if it is necessary based on the
+         *  FileSystem types that have been tested
+         */
+        assert(buf_len <= MAX_DATA_BUF_SIZE);
+
+        struct virtio_blk_outhdr hdr;
+        memcpy(&hdr, vaddr, sizeof(struct virtio_blk_outhdr));
+
+        /* Calculate the addresses to which we actually write data */
+        void *guest_buf_start = vaddr + sizeof(struct virtio_blk_outhdr);
+        void *req_status_start = vaddr + sizeof(struct virtio_blk_outhdr) + buf_len;
+
+        /* Start disk read or write chain */
+        int result = blk->driver.i_fn.raw_xfer(&blk->driver, hdr.type, hdr.sector, buf_len, (uintptr_t) guest_buf_start);
+
+        switch (result) {
+        case VIRTIO_BLK_XFER_COMPLETE:
+            *(uint8_t *)req_status_start = VIRTIO_BLK_S_OK;
+            if (VIRTIO_BLK_T_IN == hdr.type) {
+                /* We assume descriptor address at index 1 is the buffer */
+                vm_guest_write_mem(emul->vm, vaddr + sizeof(struct virtio_blk_outhdr), desc_addrs[1], buf_len);
+            }
+            /* We assume descriptor address at index 2 is the status of the IO cmd*/
+            vm_guest_write_mem(emul->vm, vaddr + sizeof(struct virtio_blk_outhdr) + buf_len, desc_addrs[2], 1);
+            complete_virtio_blk_request(emul, cookie);
+            break;
+        case VIRTIO_BLK_XFER_FAILED:
+            *(uint8_t *)req_status_start = VIRTIO_BLK_S_IOERR;
+            vm_guest_write_mem(emul->vm, vaddr + sizeof(struct virtio_blk_outhdr) + buf_len, desc_addrs[2], 1);
+            complete_virtio_blk_request(emul, cookie);
+            break;
+        }
+        /* next */
+        idx++;
+    }
+    /* update which parts of the ring we have processed */
+    emul->virtq.last_idx[emul->virtq.queue] = idx;
+}
+
+static bool emul_io_in(struct virtio_emul *emul, unsigned int offset, unsigned int size, unsigned int *result)
+{
+    bool handled = false;
+    blkif_virtio_emul_internal_t *blkif_internal = emul->internal;
+    switch (offset) {
+    case VIRTIO_PCI_HOST_FEATURES:
+        handled = true;
+        assert(size == 4);
+        *result = (BIT(VIRTIO_BLK_F_BLK_SIZE) | BIT(VIRTIO_BLK_F_SEG_MAX) | BIT(VIRTIO_BLK_F_SIZE_MAX));
+        break;
+    case VIRTIO_PCI_CONFIG_OFF(0) ... VIRTIO_PCI_CONFIG_OFF(0) + sizeof(struct virtio_blk_config):
+        handled = true;
+        assert(size == 1);
+        memcpy(result, (((uint8_t *)&blkif_internal->cfg) + offset - VIRTIO_PCI_CONFIG_OFF(0)), size);
```

Comment:
That looks a bit odd, are we assuming that the user initializes `result` to 0? Are we supposed to not touch the higher bits of `result`?

Reply:
We assume that the user initializes `result` to 0. The Linux guest should only be concerned with a single byte of memory with this emulated IO port operation so I think we're fine.